src/lib/pickles/wrap.ml

```diff
@@ -283,518 +283,391 @@ let deferred_values (type n) ~(sgs : (Backend.Tick.Curve.Affine.t, n) Vector.t)
-    let vk = Kimchi_bindings.Protocol.VerifierIndex.Fp.create index in
-    let proof = Backend.Tick.Proof.of_backend proof in
 
-    (* Specify feature flags that were used for backend proof *)
-    let actual_feature_flags =
-      let open Plonk_types.Opt.Flag in
-      { Plonk_types.Features.chacha = false
-      ; range_check = true
-      ; foreign_field_add = true
-      ; foreign_field_mul = false
-      ; xor = false
-      ; rot = false
-      ; lookup = true
-      ; runtime_tables = false
-      }
-    in
-
-    (* Run the recursive proof test with supplied feature flags *)
-    run_custom_gate_tests actual_feature_flags vk proof
-  with _e ->
-    Printexc.print_backtrace stdout ;
-    Out_channel.flush stdout ;
-    exit 2
+    type example =
+         Kimchi_bindings.Protocol.SRS.Fp.t
+      -> Kimchi_bindings.Protocol.Index.Fp.t
+         * Pasta_bindings.Fp.t list
+         * ( Pasta_bindings.Fq.t Kimchi_types.or_infinity
+           , Pasta_bindings.Fp.t )
+           Kimchi_types.prover_proof
+
+    module type SETUP = sig
+      val example : example
+
+      (* Feature flags tused for backend proof *)
+      val actual_feature_flags : bool Plonk_types.Features.t
+    end
+
+    (* [Make] is the test functor.
+
+       Given a test setup, compute different test configurations and define 3
+       test for said configurations. *)
+    module Make (S : SETUP) = struct
+      (* Generate foreign field multiplication test backend proof using Kimchi,
+         obtaining the proof and corresponding prover index.
+
+         Note: we only want to pay the cost of generating this proof once and
+         then reuse it many times for the different recursive proof tests. *)
+      let index, public_input, proof = S.example srs
+
+      (* Obtain verifier key from prover index and convert backend proof to
+         snarky proof *)
+      let vk = Kimchi_bindings.Protocol.VerifierIndex.Fp.create index
+
+      let proof = Backend.Tick.Proof.of_backend proof
+
+      let test_feature_flags_configs =
+        generate_test_feature_flag_configs S.actual_feature_flags
+
+      let runtest feature_flags =
+        run_recursive_proof_test S.actual_feature_flags feature_flags
+          public_input vk proof
+
+      let%test "true -> yes" = runtest test_feature_flags_configs.true_is_yes
+
+      let%test "true -> maybe" =
+        runtest test_feature_flags_configs.true_is_maybe
+
+      let%test "all maybes" = runtest test_feature_flags_configs.all_maybes
+    end
+
+    (* Small combinators to lift gate example signatures to the expected
+       signatures for the tests. This amounts to generating the list of public
+       inputs from either no public inputs, a single one or a pair of inputs
+       returned by the gate example. *)
+
+    let no_public_input gate_example srs =
+      let index, proof = gate_example srs in
+      (index, [], proof)
+
+    let public_input_1 gate_example srs =
```

Comment:
perhaps you would prefer that rust tests which do not use public inputs return an empty list perhaps? instead of duplicating these two parts here

Reply:
meaning, cannot we return Vec\<CamlFp\> and return empty vec![] if we do not have public inputs?

Reply:
It's another possibility, indeed. I thought it easier to do it on the OCaml side ;-)